### ISSUE TITLE
Is multi-gpu working with cuda-available devices?

### ISSUE THREAD

**Issue by anselmoferreira** [keras-team/keras#9229]
Please make sure that the boxes below are checked before you submit your issue. If your issue is an implementation question, please ask your question on [StackOverflow](http://stackoverflow.com/questions/tagged/keras) or [join the Keras Slack channel](https://keras-slack-autojoin.herokuapp.com/) and ask there instead of filing a GitHub issue.

Thank you!

- [x ] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/keras-team/keras.git --upgrade --no-deps

- [ x] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [ x] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [ x]  Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

I have a strange problem and I want a help from here, I have a computer with 8 GPUs and I would like to use only 4 of them in the multi gpu training. Here is my code:

from keras.preprocessing.image import ImageDataGenerator
from keras.utils import np_utils
from keras.callbacks import ReduceLROnPlateau, CSVLogger, EarlyStopping, ModelCheckpoint
from generator_aug import loaddata_generator
import numpy as np
from keras.applications.xception import Xception
import os
import tensorflow as tf
from keras.models import Model
from keras.layers import Dense, GlobalAveragePooling2D, BatchNormalization, Dropout
os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
os.environ["CUDA_VISIBLE_DEVICES"] = "0,1,2,3"
from keras.utils.training_utils import multi_gpu_model 

lr_reducer = ReduceLROnPlateau(factor=np.sqrt(0.1), cooldown=0, patience=5, min_lr=0.5e-6)
early_stopper = EarlyStopping(min_delta=0.001, patience=20)
csv_logger = CSVLogger('corrected_jpeg_xception2.csv')
model_names = 'trained_model/corrected2_jpeg_xception_' + '{epoch:02d}-{val_acc:.2f}.hdf5'
model_checkpoint = ModelCheckpoint(model_names, 'val_loss', verbose=1,
                                                save_best_only=True)

								
batch_size = 128
nb_classes = 10
nb_epoch = 200
data_augmentation = True

# input image dimensions
img_rows, img_cols = 299, 299
# The CIFAR10 images are RGB.
img_channels = 3

with tf.device("/cpu:0"):
    base_model = Xception(weights=None,include_top=False)
    base_model.load_weights('xception_weights_tf_dim_ordering_tf_kernels_notop.h5')
    x = base_model.output
    x = GlobalAveragePooling2D()(x)
    x=Dropout(0.2)(x)
    x = Dense(256, activation='relu')(x)
    predictions = Dense(10, activation='softmax')(x)
    # this is the model we will train
    model = Model(inputs=base_model.input, outputs=predictions)
    print(model.summary())
    #model.load_weights('/home/spscup/cameraeye/large_dataset/part2_jpeg_pre_xception/trained_model/corrected_jpeg_xception_10-0.99.hdf5')

model = multi_gpu_model(model, gpus=4)

model.compile(loss='categorical_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

if not data_augmentation:
    print('Not using data augmentation.')
    model.fit(x_train, y_train,
              batch_size=batch_size,
              nb_epoch=nb_epoch,
              validation_data=(x_validation, y_validation),
              shuffle=True,
              callbacks=[lr_reducer, early_stopper, csv_logger])
else:
    print('Using real-time data augmentation.')
    # This will do preprocessing and realtime data augmentation:
    datagen = ImageDataGenerator(
        featurewise_center=False,  # set input mean to 0 over the dataset
        samplewise_center=False,  # set each sample mean to 0
        featurewise_std_normalization=False,  # divide inputs by std of the dataset
        samplewise_std_normalization=False,  # divide each input by its std
        zca_whitening=False,  # apply ZCA whitening
        rotation_range=0,  # randomly rotate images in the range (degrees, 0 to 180)
        width_shift_range=0.1,  # randomly shift images horizontally (fraction of total width)
        height_shift_range=0.1,  # randomly shift images vertically (fraction of total height)
        horizontal_flip=True,  # randomly flip images
        vertical_flip=False)  # randomly flip images

    # Compute quantities required for featurewise normalization
    # (std, mean, and principal components if ZCA whitening is applied).
    #datagen.fit(x_train)

    # Fit the model on the batches generated by datagen.flow().
    model.fit_generator( generator=loaddata_generator('train_merge_JPG_images_random.txt',batch_size,datagen),
              steps_per_epoch=2801,
              epochs=nb_epoch,
              validation_data=loaddata_generator('validation_merge_JPG_images_random.txt',batch_size),
              callbacks=[lr_reducer, early_stopper, csv_logger,model_checkpoint],validation_steps=1200)

I am using os.environ["CUDA_VISIBLE_DEVICES"] = "0,1,2,3" so I can use the multi gpu training on only these four GPUs. I can see that only for GPUs are used to process my cnn after the nvidia-smi command but the same PID is running on all 8 and the memory of all GPUs are used.

Let me ask you something. Is the multi-gpu call on keras forced to use all GPUs of my computer? Can I select only some of them to multi-gpu processing?

**Comment by tjochmann** [comment]
> Let me ask you something. Is the multi-gpu call on keras forced to use all GPUs of my computer? Can I select only some of them to multi-gpu processing?

Yes, you can. Since Keras 2.1.2, the gpus argument can be either a single integer with the number of GPUs or a list of GPU ids.
